[9.0-preview.4] Failed tests not reported to the console in CI on MacOS

### Describe the bug

Using the .NET 9 preview 4 SDK, if a test fails in GitHub Actions CI where [GitHubActionsTestLogger](https://github.com/Tyrrrz/GitHubActionsTestLogger) is used, no information about the failed test is printed to the console.

```console
Running tests...
  Determining projects to restore...
  Restored /home/runner/work/xunit-logging/xunit-logging/tests/SampleApp/SampleApp.csproj (in 321 ms).
  Restored /home/runner/work/xunit-logging/xunit-logging/tests/Logging.XUnit.Tests/MartinCostello.Logging.XUnit.Tests.csproj (in 3.52 sec).
  1 of 3 projects are up-to-date for restore.
/usr/share/dotnet/sdk/9.0.100-preview.4.24267.66/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/home/runner/work/xunit-logging/xunit-logging/tests/Logging.XUnit.Tests/MartinCostello.Logging.XUnit.Tests.csproj]
/usr/share/dotnet/sdk/9.0.100-preview.4.24267.66/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/home/runner/work/xunit-logging/xunit-logging/src/Logging.XUnit/MartinCostello.Logging.XUnit.csproj::TargetFramework=net8.0]
  MartinCostello.Logging.XUnit -> /home/runner/work/xunit-logging/xunit-logging/artifacts/bin/MartinCostello.Logging.XUnit/release_net8.0/MartinCostello.Logging.XUnit.dll
/usr/share/dotnet/sdk/9.0.100-preview.4.24267.66/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/home/runner/work/xunit-logging/xunit-logging/tests/SampleApp/SampleApp.csproj]
  SampleApp -> /home/runner/work/xunit-logging/xunit-logging/artifacts/bin/SampleApp/release/SampleApp.dll
  MartinCostello.Logging.XUnit.Tests -> /home/runner/work/xunit-logging/xunit-logging/artifacts/bin/MartinCostello.Logging.XUnit.Tests/release/MartinCostello.Logging.XUnit.Tests.dll
  [coverlet] _mapping file name: 'CoverletSourceRootsMapping_MartinCostello.Logging.XUnit.Tests'
/usr/share/dotnet/sdk/9.0.100-preview.4.24267.66/Microsoft.TestPlatform.targets(46,5): error MSB6006: "dotnet" exited with code 1. [/home/runner/work/xunit-logging/xunit-logging/tests/Logging.XUnit.Tests/MartinCostello.Logging.XUnit.Tests.csproj]
```

I noticed this in a previous issue, but have only just come back around to check in on it: https://github.com/dotnet/sdk/issues/39665#issuecomment-2013172418

The only way to see which test is failing is to look at the [workflow summary](https://github.com/martincostello/xunit-logging/actions/runs/9316133081), which shows which test failed for Linux and Windows, but interestingly does _not_ show anything [for macOS](https://github.com/martincostello/xunit-logging/actions/runs/9316133081#summary-25643869775).

### macOS

![image](https://github.com/dotnet/sdk/assets/1439341/d8b714b0-234a-4801-a593-20f33f5c86af)

### Linux

![image](https://github.com/dotnet/sdk/assets/1439341/86ec0f3e-2228-4bfb-8cbe-1c43cf74dc36)

Contrast this to the non-CI behaviour, where the test failure is obvious:

```console
Running tests...
Restore complete (1.0s)
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  MartinCostello.Logging.XUnit net8.0 succeeded (0.2s) → artifacts\bin\MartinCostello.Logging.XUnit\release_net8.0\MartinCostello.Logging.XUnit.dll
  SampleApp succeeded (4.3s) → artifacts\bin\SampleApp\release\SampleApp.dll
  MartinCostello.Logging.XUnit.Tests succeeded (3.7s) → artifacts\bin\MartinCostello.Logging.XUnit.Tests\release\MartinCostello.Logging.XUnit.Tests.dll
  MartinCostello.Logging.XUnit.Tests test failed with 2 error(s) and 2 warning(s) (5.4s)
    C:\Program Files\dotnet\sdk\9.0.100-preview.4.24267.66\Microsoft.TestPlatform.targets(46,5): warning :
      [xUnit.net 00:00:00.51] MartinCostello.Logging.XUnit.Tests: [2024-05-31 09:31:44Z] info: MartinCostello.Loggi
      ng.XUnit.Integration.DatabaseFixture[0]
            Initializing database
    C:\Program Files\dotnet\sdk\9.0.100-preview.4.24267.66\Microsoft.TestPlatform.targets(46,5): warning :
      [xUnit.net 00:00:00.54] MartinCostello.Logging.XUnit.Tests: [2024-05-31 09:31:44Z] info: MartinCostello.Loggi
      ng.XUnit.Integration.DatabaseFixture[0]
            Disposing database
    C:\Program Files\dotnet\sdk\9.0.100-preview.4.24267.66\Microsoft.TestPlatform.targets(46,5): error : [xUnit.net 00:00:00.73]     Calculator_Sums_Two_Equal_Integers [FAIL]
    C:\Coding\martincostello\xunit-logging\tests\Logging.XUnit.Tests\Examples.cs(21): error VSTEST1:
      Calculator_Sums_Two_Equal_Integers: MartinCostello.Logging.XUnit.Examples.Calculator_Sums_Two_Equal_Integers(
      ) Shouldly.ShouldAssertException : actual
          should not be
      4
          but was

Test summary: total: 70, failed: 1, succeeded: 69, skipped: 0, duration: 4.1s
Build failed with 2 error(s) and 2 warning(s) in 16.4s
```

### To Reproduce

Build https://github.com/martincostello/xunit-logging/pull/662/commits/dc54d8e3c64b9a0bc3c1a42f5045b9d572cfcbaa in GitHub Actions.

### Exceptions (if any)

None.

### Further technical details

.NET SDK `9.0.100-preview.4.24267.66`

/cc @nohwnd

--- a/src/Microsoft.TestPlatform.Build/Tasks/VSTestTask2.cs
+++ b/src/Microsoft.TestPlatform.Build/Tasks/VSTestTask2.cs
@@ -49,8 +49,10 @@ public class VSTestTask2 : ToolTask, ITestTask
 
     private readonly string _messageSplitter = "||||";
     private readonly string[] _messageSplitterArray = new[] { "||||" };
+    private readonly string _ansiReset = "\x1b[39;49m";
 
     private readonly bool _disableUtf8ConsoleEncoding;
+    private readonly bool _canBePrependedWithAnsi;
 
     protected override string? ToolName => RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "dotnet.exe" : "dotnet";
 
@@ -58,6 +60,9 @@ public class VSTestTask2 : ToolTask, ITestTask
     {
         // Unless user opted out, use UTF encoding, which we force in vstest.console.
         _disableUtf8ConsoleEncoding = Environment.GetEnvironmentVariable("VSTEST_DISABLE_UTF8_CONSOLE_ENCODING") == "1";
+        var isPrependedWithAnsi = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_CONSOLE_ALLOW_ANSI_COLOR_REDIRECTION");
+        // On macOS and Linux the messages are prepended with ANSI reset sequence.
+        _canBePrependedWithAnsi = isPrependedWithAnsi?.ToLowerInvariant() == "true" || isPrependedWithAnsi == "1";
         LogStandardErrorAsError = false;
         StandardOutputImportance = "High";
     }
@@ -66,6 +71,17 @@ public class VSTestTask2 : ToolTask, ITestTask
     {
         var useTerminalLogger = true;
         Debug.WriteLine($"VSTESTTASK2: Received output {singleLine}, importance {messageImportance}");
+        bool wasPrependedWithAnsi = false;
+
+        if (_canBePrependedWithAnsi)
+        {
+            while (singleLine.StartsWith(_ansiReset))
+            {
+                wasPrependedWithAnsi = true;
+                singleLine = singleLine.Substring(_ansiReset.Length);
+            }
+        }
+
         if (TryGetMessage(singleLine, out string name, out string?[] data))
         {
             // See MSBuildLogger.cs for the messages produced.
@@ -75,7 +91,9 @@ public class VSTestTask2 : ToolTask, ITestTask
                 case "output-info":
                     if (data[0] != null)
                     {
-                        LogMSBuildOutputMessage(data[0]!);
+                        // This is console output was prepended with ANSI reset, add it back.
+                        string info = wasPrependedWithAnsi ? _ansiReset + data[0]! : data[0]!;
+                        LogMSBuildOutputMessage(info);
                     }
                     break;
                 case "output-warning":

